josh-core/src/history.rs

```diff
@@ -205,40 +205,49 @@ pub fn rewrite_commit(
     rewrite_data: RewriteData,
     unsign: bool,
 ) -> JoshResult<git2::Oid> {
-    let message = rewrite_data
-        .message
-        .unwrap_or(base.message_raw().unwrap_or("no message").to_string());
-    let tree = &rewrite_data.tree;
-
-    let a = base.author();
-    let new_a = if let Some((author, email)) = rewrite_data.author {
-        git2::Signature::new(&author, &email, &a.when())?
-    } else {
-        a
-    };
+    let odb = repo.odb()?;
+    let odb_commit = odb.read(base.id())?;
+    assert!(odb_commit.kind() == git2::ObjectType::Commit);
+
+    // gix_object uses byte strings for Oids, but in hex representation, not raw bytes. Its `Format` implementation
+    // writes out hex-encoded bytes. Because CommitRef's reference lifetimes we have to this, before creating CommitRef
+    let tree_id = format!("{}", rewrite_data.tree.id());
+    let parent_ids = parents
+        .iter()
+        .map(|x| format!("{}", x.id()))
+        .collect::<Vec<_>>();
 
-    let c = base.committer();
-    let new_c = if let Some((committer, email)) = rewrite_data.committer {
-        git2::Signature::new(&committer, &email, &c.when())?
-    } else {
-        c
-    };
+    let mut commit = gix_object::CommitRef::from_bytes(odb_commit.data())?;
```

Comment:
The idea here would be to call `.to_owned()` so dealing with changing the commit is a little less cumbersome, hopefully.

You might also find [this method](https://github.com/Byron/gitoxide/blob/abcfb659786425ec09eff6b644cd2ad36b7d6bc4/gix-object/src/commit/ref_iter.rs#L57), as it's related to handling signatures, even though I am not sure it perfectly applies here.

Reply:
Thanks for stopping by @Byron!

> The idea here would be to call `.to_owned()` so dealing with changing the commit is a little less cumbersome, hopefully.

I had wondered about this. I opted to stay with the ref struct once I had it working but agree the Owned version might be easier. I might do that and compare, but I'm pretty happy with how it is now.

The thing I found the most cumbersome was not the lifetimes but he handling of IDs as hex, actually. Because it was a byte slice and taking bytes, it assumed this would be 20 raw bytes, as per `git2` and was surprised that this was 40-char hex bytes. I think being able to more naturally work with both would be an ergonomic improvement.

> You might also find [this method](https://github.com/Byron/gitoxide/blob/abcfb659786425ec09eff6b644cd2ad36b7d6bc4/gix-object/src/commit/ref_iter.rs#L57), as it's related to handling signatures, even though I am not sure it perfectly applies here.

Because JOSH is rewriting commits in ways that will break the signature, the choice indicated by `unsign` is merely whether to remove the signature altogether or to leave it broken (so that the commit can be inversely rewritten). Validating it or using the signature itself doesn't seem applicable (in this area).

But that method does look convenient, so making a mental note!

Reply:
You are welcome!

> The thing I found the most cumbersome was not the lifetimes but he handling of IDs as hex, actually. Because it was a byte slice and taking bytes, it assumed this would be 20 raw bytes, as per `git2` and was surprised that this was 40-char hex bytes. I think being able to more naturally work with both would be an ergonomic improvement.

The refs represent what's in the buffer, and the buffer contains hex-hashes. The fully owned version is at liberty to more naturally represent the data, so it would use `gix_hash::ObjectId` instead.

> But that method does look convenient, so making a mental note!

I think with the signature range (i.e. where the signature is in the bytes buffer) one could easily build an unsigned version of that buffer. That one could parse and alter, while leaving the signature out of there which I thought was happening here.

Admittedly I didn't look closely enough to truly understand what's happening, so this is merely a dump of what I know and you can mix and match :).